### PR TITLE
8315968: Move java.util.Digits to jdk.internal.util and refactor to reduce duplication

### PR DESCRIPTION
**Pull request by wenshao**
java.util.DecimalDigits::DIGITS and java.lang.StringLatin1.PACKED_DIGITS are duplicates, We need to move java.util.Digits/OctalDigits/DecimalDigits/HexDigits to the jdk.internal.util package, and modify these classes to public class, so that classes in other packages can also access them.

DecimalDigits::DIGITS provides a new digitPair static method, used to replace StringLatin1.PACKED_DIGITS access.

In order to be consistent with the original StringLatin1.PACKED_DIGITS, OctalDigits::DIGITS and DecimalDigits::DIGITS are modified to little-endian. HexDigits::DIGITS will also be modified after PR #14745 is merged.

These changes will be used by PR #15658 #15555

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315968](https://bugs.openjdk.org/browse/JDK-8315968): Move java.util.Digits to jdk.internal.util and refactor to reduce duplication (**Enhancement** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author) ⚠️ Review applies to [27174f45](https://git.openjdk.org/jdk/pull/15651/files/27174f45ba46f3f4b9055419110125f75edb159f)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15651/head:pull/15651` \
`$ git checkout pull/15651`

Update a local copy of the PR: \
`$ git checkout pull/15651` \
`$ git pull https://git.openjdk.org/jdk.git pull/15651/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15651`

View PR using the GUI difftool: \
`$ git pr show -t 15651`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15651.diff">https://git.openjdk.org/jdk/pull/15651.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15651#issuecomment-1712856174)

### CODE REVIEW

**Comment by RogerRiggs** [comment]
Instead of packing more stuff into SharedSecrets, how about moving some common utilities into package jdk.internal.util or a new package jdk.internal.string.
For example, Digits, HexDigits, DecimalDigit, OctalDigit.
That would make access better without the overhead shared secrets.

**Comment by cl4es** [comment]
Yes, moving java.util.Digits to jdk.internal.util was what I had in mind, too. SharedSecrets are only necessary when we have things that need to stay package-private, such as member functions on String. Static utilities like this are generally easier to maintain when they are public but in a non-exported package like jdk.internal.util.

File: src/java.base/share/classes/java/lang/Long.java
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/Long.java:45]: Please don't use static import for methods. It makes it much harder to find where they come from.

File: src/java.base/share/classes/java/util/Digits.java
Comment (RogerRiggs) [inline src/java.base/share/classes/java/util/Digits.java:36]: Don't break up the trio, move all three classes and the interface to jdk.internal.util.
I don't see the value in the INSTANCE values but keep it intact.

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
This would cleaner and easier to review, if you moved the classes together and did not refactor the functions. Only include the dependencies. Update the functions in a separate PR.  (We can retitle the issue to match).

File: src/java.base/share/classes/java/util/Digits.java
Comment (wenshao) [reply src/java.base/share/classes/java/util/Digits.java:36]: HexDigits I will change after the PR #1745 is merged,  Can you add comment /sponsor for me?

Comment (wenshao) [reply src/java.base/share/classes/java/util/Digits.java:36]: I want to remove the Digits interface and use static methods

Comment (cl4es) [reply src/java.base/share/classes/java/util/Digits.java]: I agree with @RogerRiggs that these should be moved together and with as few changes as possible. We can do redesigns in follow-ups.

I'd be OK with adding static variants of each method as needed, leaving the instance methods unchanged. 

I'll note that these `java.util.Digits` came in as part of what's currently a preview feature (https://openjdk.org/jeps/430) and the author (@JimLaskey) might have plans that requires an implementation that could be passed around in the final version - so design changes should be coordinated with him.

File: src/java.base/share/classes/java/lang/Integer.java
Comment (y1yang0) [inline src/java.base/share/classes/java/lang/Integer.java:459]: Would better to update related comments as well

https://github.com/openjdk/jdk/blob/d2e11593006dc32fb8ebbaf12488b8758c8a19ee/src/hotspot/share/opto/stringopts.cpp#L1163-L1165

There are several occurrences in stringopt phase, grep to find them.

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:201]: It's unused now. Do we expect usages to be added in following PRs?

Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:201]: These weren't in `java.util.DecimalDigits` but have been copied from `java.lang.StringLatin1` - part of an unfinished refactoring? There's no clear-cut answer where these best fits but it seems reasonable to keep them in `StringLatin1` and `StringUTF16` respectively.

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:201]: > It's unused now. Do we expect usages to be added in following PRs?

I plan to provide the getChars method in this PR and use it in my other PR #15555. I also plan to submit another PR to optimize java.time toString, and I will also use this method. https://github.com/wenshao/jdk/commit/cc82e8c11aeb27c0f8581363f7133c85f6022ca1

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java]: > These weren't in `java.util.DecimalDigits` but have been copied from `java.lang.StringLatin1` - part of an unfinished refactoring? There's no clear-cut answer where these best fits but it seems reasonable to keep them in `StringLatin1` and `StringUTF16` respectively.

StringLatin1.getChars has been moved to DecimalDigits.getCharsLatin1

File: src/java.base/share/classes/jdk/internal/util/OctalDigits.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/util/OctalDigits.java:68]: Can you revert these IDE reformatting changes? So that we don't pollute the changeset and the Git blame log, making reviews and reverting patches easier.

For example, only 3 lines of changes are truly necessary in this class:
Changing the package, `public ` before `final class OctalDigits` and `static final Digits INSTANCE`.

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:143]: I suggest splitting the moves of `stringSize` `getChars` into a new PR dependent on this one; your future date and time optimizations can depend on that one, which exposes stringSize.

Having the DecimalDigits package move and `stringSize` `getChars` moves together complicates the file changes, and it's hard to detect if there's any accidental typo/malicious code in the new additions.

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:45]:
```suggestion
```
Per previous comments.

Remember to add another line of 
```java
import jdk.internal.util.DecimalDigits;
```
 above!

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:1549]:
```suggestion
            int packed = (int) DecimalDigits.digitPair(r);
```

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:1558]:
```suggestion
            int packed = (int) DecimalDigits.digitPair(-i);
```

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:1596]:
```suggestion
            int packed = (int) DecimalDigits.digitPair((int)((q * 100) - i));
```

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:1610]:
```suggestion
            int packed = (int) DecimalDigits.digitPair((q2 * 100) - i2);
```

File: src/java.base/share/classes/java/lang/StringUTF16.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringUTF16.java:1622]:
```suggestion
            int packed = (int) DecimalDigits.digitPair(-i2);
```

File: src/java.base/share/classes/java/lang/StringLatin1.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringLatin1.java:35]:
```suggestion
import jdk.internal.util.ArraysSupport;
import jdk.internal.util.DecimalDigits;
```
Assuming you will restore the `getChars` here.

File: src/java.base/share/classes/java/lang/StringLatin1.java
Comment (liach) [inline src/java.base/share/classes/java/lang/StringLatin1.java:42]:
```suggestion
```
Redundant.

**Review by liach — COMMENTED** [review]
In short: I suggest splitting the change of moving `getChars` `stringSize` into another patch; they are sufficiently distinct and can shrink this PR's size by a half at least.

总结：推荐这个PR专门移除`StringLatin1.PACKED_DIGITS`和`Digits`移包。
迁移`getChars`和`stringSize`分另外一个补丁；这两个内容唯一共通点是`StringLatin1.getChars`有些冲突，但是除此之外没必要两个补丁合并，大幅增加代码改动，难review，回滚和git blame，比如这个补丁里面Long AbstractStringBuilder等一堆改动实际上是这两个method，和`PACKED_DIGITS`无关。

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:145]: If this PR is split into two PRs, the other two PRs I submitted #15658 #15555 cannot be based on this PR.

File: src/java.base/share/classes/jdk/internal/util/OctalDigits.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/util/OctalDigits.java:37]: Can you restore the order of methods here to minimize the number of lines of code are touched?

File: src/java.base/share/classes/jdk/internal/util/OctalDigits.java
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/util/OctalDigits.java:41]: The constructor should be after the field definitions. 
Don't let some tool reformat the code and move things around.

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java
Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:1]: I just looked it up and git actually doesn't have commands for renames like hg, but rather decides if something is a delete+add or a rename based on the amount of changes. Too many changes and it'll automatically look like a remove+add. 

Which is yet another argument for splitting up this in multiple PRs. One to move, one to refactor.

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:1]: Refactoring of stringSize has been restored, But the problem of 'delete & add' is still not solved, I tried rebase but it couldn't be solved.

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java]: Refactoring of stringSize has been restored

**Review by liach — APPROVED** [review]

**Comment by cl4es** [comment]
Running some additional testing. This mostly looks fine. 

One issue is that you're swapping the byte-order in `DecimalDigits::DIGITS` but not in `OctalDigits` and `HexDigits`. I think we need to keep these internally consistent to avoid surprises.

I also would like to see performance numbers of the byte order swap evaluated in isolation. I suspect the real effect is small and might be due to JIT noise rather than a real effect, and that this swap got rushed in without solid evidence that it helps. 

If there's no significant performance difference I would prefer if we kept `DecimalDigits::DIGITS` big-endian encoded - which is more intuitive to most - and adjust code depending on `DecimalDigits::digitPair` to use `ByteArray` rather than `ByteArrayLittleEndian`.

**Comment by cl4es** [comment]
> If there's no significant performance difference I would prefer if we kept `DecimalDigits::DIGITS` big-endian encoded - which is more intuitive to most - and adjust code depending on `DecimalDigits::digitPair` to use `ByteArray` rather than `ByteArrayLittleEndian`.

Just a datapoint but when I test implementing `HexFormat::toHexDigits` using either `ByteArray` or `ByteArrayLittleEndian` on the difference is in the noise:

```
Name                           Cnt  Base   Error   Test   Error  Unit   Diff%
HexFormatBench.toHexDigitsLong  15 1,950 ± 0,012  1,941 ± 0,016 us/op    0,5% (p = 0,081 )
```

**Comment by wenshao** [comment]
> Running some additional testing. This mostly looks fine.
> 
> One issue is that you're swapping the byte-order in `DecimalDigits::DIGITS` but not in `OctalDigits` and `HexDigits`. I think we need to keep these internally consistent to avoid surprises.
> 
> I also would like to see performance numbers of the byte order swap evaluated in isolation. I suspect the real effect is small and might be due to JIT noise rather than a real effect, and that this swap got rushed in without solid evidence that it helps.
> 
> If there's no significant performance difference I would prefer if we kept `DecimalDigits::DIGITS` big-endian encoded - which is more intuitive to most - and adjust code depending on `DecimalDigits::digitPair` to use `ByteArray` rather than `ByteArrayLittleEndian`.

I prefer to use little endian, after all, most environments are little endian. Changes to HexDigit will have to wait until #PR 14745 is merged.

**Comment by RogerRiggs** [comment]
Please update this PR title and description to indicate this refactoring to move to jdk.internal.util.
I can update the Jira title and description to match after that.

**Comment by wenshao** [comment]
> Please update this PR title and description to indicate this refactoring to move to jdk.internal.util. I can update the Jira title and description to match after that.

The title has been updated, please help update the title of JIRA

**Comment by RogerRiggs** [comment]
> The title has been updated, please help update the title of JIRA

The description needs an update too.

**Comment by liach** [comment]
I have updated the JBS bug title and explicitly stated the actions in this patch:
1. Move Digits to jdk.internal.util
2. Keep the tables in DecimalDigits and access via a new digitPair static method
3. Change the DecimalDigits' table to be LE (instead of BE) and update implementation

**Comment by wenshao** [comment]
> > The title has been updated, please help update the title of JIRA
> 
> The description needs an update too.

The description has also been updated

**Comment by cl4es** [comment]
> I prefer to use little endian, most environments are little endian. Changes to HexDigit will have to wait until #PR 14745 is merged.

I don't have a very strong opinion except that we should be consistent across these related implementations. If you could either include `OctalDigits` changes in #14745 or prepare a separate PR for it I'm OK with this.

**Review by RogerRiggs — APPROVED** [review]
Thanks for the updates.

**Pull request closed without merging**
